### PR TITLE
tests for workbook saver

### PR DESCRIPTION
**Pull request by mykola-pysarev**

### CODE REVIEW

File: xmind/tests/test_workbook_saver.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        self._workbook = MagicMock()
+        self._test_el = WorkbookSaver(self._workbook)
+
+    def getLogger(self):
+        if not getattr(self, '_logger', None):
+            self._logger = logging.getLogger('WoorkbookSaverTest')
+        return self._logger
+
+    def test_excessive_parameters(self):
+        # _workbook = MagicMock()
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: No reason to keep commented out code

File: xmind/tests/test_workbook_saver.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        # _element = WorkbookSaver(_workbook)
+        _parameters = [
+            ('_get_content', 0),
+          #  ('save', 1)
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: you can use here ```('save', (1, False))``` and avoid block below

File: xmind/tests/test_workbook_saver.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+        self.getLogger().debug("Got an exception: '%s'", _ex.exception.args[0])
+        self.assertTrue(_ex.exception.args[0].find(
+            "%s() takes" % 'save') != -1)
+
+        for pair in _parameters:
+            with self.subTest(pair=pair):
+                self._test_method_by_excessive_parameters(pair, self._test_el)
+
+    def test_WoorkbookSaver_inint(self):
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: mistypo ```_inint``` -> ```_init```

Comment (andrii-z4i) [reply xmind/tests/test_workbook_saver.py]: btw, I think you can use approach which @vlad-shulika proposed to put 
```
_parameters = [
            ('_get_content', 0),
            ('save', (1, False)),
            ('__init__', 1)
        ]
``` 
into ```test_excessive_parameters``` and it will cover ```__init__``` as well

File: xmind/tests/test_workbook_saver.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        self.assertTrue(_ex2.exception.args[0].find(
+                "__init__() takes") != -1)
+
+        # Call with one parameter
+        _test_object = WorkbookSaver('workbook')
+        self.assertEqual(_test_object._workbook,'workbook')
+
+    def test_get_content(self):
+
+        with patch('xmind.utils.join_path') as _utils_join_path:
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: It's very difficult to read so long hierarchy of ```with``` 
I would recommend to go with patching of each method before actual call of it.

File: xmind/tests/test_workbook_saver.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                        _el = self._test_el._get_content()
+
+        _utils_temp_dir.assert_called_once_with()
+        _utils_join_path.assert_called_once_with('temp dir', CONTENT_XML)
+        _m.assert_called_once_with('joined path', 'w', encoding="utf-8")
+        _output.assert_called_once_with(_m())
+        self.assertEqual(_el, 'joined path')
+
+    def test_save_with_no_path_no_get_path(self):
+        with patch.object(self._workbook, 'get_path') as _get_path:
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: what is the reason to patch ```_workbook```? it's already ```MagicMock``` you can just define
```self._workbook.get_path.return_value = None```
in your test

File: xmind/tests/test_workbook_saver.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                self._test_el.save('path')
+
+        self.assertTrue(_ex.exception.args[0].find("stop point") == 0)
+        _get_abs_path.assert_called_once_with('path')
+
+    def test_save_no_path_is_get_path(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
+            _get_abs_path.side_effect = Exception('stop point')
+
+            with patch.object(self._workbook, 'get_path') as _get_path:
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: Again, ```self._workbook``` itself is ```MagicMock```

File: xmind/tests/test_workbook_saver.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                _get_path.return_value = 'get path'
+                with self.assertRaises(Exception) as _ex:
+                    self._test_el.save()
+
+        self.assertTrue(_ex.exception.args[0].find("stop point") == 0)
+        _get_abs_path.assert_called_once_with('get path')
+        _get_path.assert_called_once_with()
+
+    def test_save_if_ext_is_not_equal_xmind_ext(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: More than 2 levels of nesting is hard to follow

File: xmind/tests/test_workbook_saver.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                _split_ext.return_value = 'name', 'ext'
+                with self.assertRaises(Exception) as _ex:
+                    self._test_el.save('path')
+
+        self.assertTrue(_ex.exception.args[0].find("XMind filenames require a '%s' extension" % XMIND_EXT) != -1)
+        _get_abs_path.assert_called_once_with('path')
+        _split_ext.assert_called_once_with('abs path')
+
+    def test_save_if_ext_is_equal_xmind_ext(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
```
Comment (andrii-z4i) [inline xmind/tests/test_workbook_saver.py]: More than 2 levels of nesting is hard to follow

**Review by andrii-z4i — CHANGES_REQUESTED** [review]
Very good progress!!!
====

please address comments and go to master 😄

File: xmind/tests/test_workbook_saver.py (hunk 1 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: fixed

File: xmind/tests/test_workbook_saver.py (hunk 2 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: done

File: xmind/tests/test_workbook_saver.py (hunk 3 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

File: xmind/tests/test_workbook_saver.py (hunk 4 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

File: xmind/tests/test_workbook_saver.py (hunk 5 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

File: xmind/tests/test_workbook_saver.py (hunk 6 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

File: xmind/tests/test_workbook_saver.py (hunk 7 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

File: xmind/tests/test_workbook_saver.py (hunk 8 above)
Comment (mykola-pysarev) [reply xmind/tests/test_workbook_saver.py]: Done

**Review by andrii-z4i — APPROVED** [review]
Thanks for addressing all comments!
